cpp/src/column/column_device_view.cu

```diff
@@ -45,51 +44,72 @@ column_device_view::column_device_view( column_view source, ptrdiff_t h_ptr, ptr
 {
-  if( count_descendants(source) > _num_children ) {
-    CUDF_FAIL("Columns with grand-children are not currently supported.");
-  }
-  if( _num_children > 0 )
+  size_type num_children = source.num_children();
+  if( num_children > 0 )
   {
     column_device_view* h_column = reinterpret_cast<column_device_view*>(h_ptr);
     column_device_view* d_column = reinterpret_cast<column_device_view*>(d_ptr);
-    d_children = d_column;
+    int8_t* h_end = (int8_t*)(h_column + num_children);
+    int8_t* d_end = (int8_t*)(d_column + num_children);
+    d_children = d_column; // set member ptr to device memory
     for( size_type idx=0; idx < _num_children; ++idx )
-    { // inplace-new each child
+    { // inplace-new each child into host memory
       column_view child = source.child(idx);
-      new(h_column) column_device_view(child);
-      h_column++;
+      new(h_column) column_device_view(child,(ptrdiff_t)h_end,(ptrdiff_t)d_end);
+      h_column++; // adv to next child
+      // update the pointers for holding this child column's child data
+      auto col_child_data_size = extent(child) - sizeof(child);
+      h_end += col_child_data_size;
+      d_end += col_child_data_size;
     }
   }
 }
 
 // Construct a unique_ptr that invokes `destroy()` as it's deleter
 std::unique_ptr<column_device_view, std::function<void(column_device_view*)>> column_device_view::create(column_view source, cudaStream_t stream) {
   size_type num_children = source.num_children();
-  if( count_descendants(source) > num_children ) {
-    CUDF_FAIL("Columns with grand-children are not currently supported.");
-  }
   auto deleter = [](column_device_view* v) { v->destroy(); };
   std::unique_ptr<column_device_view, decltype(deleter)> p{
       new column_device_view(source), deleter};
 
   if( num_children > 0 )
   {
-    // create device memory for the children
-    RMM_TRY(RMM_ALLOC(&p->d_children, sizeof(column_device_view)*num_children, stream));
-    // build the children into CPU memory first
-    std::vector<uint8_t> buffer(sizeof(column_device_view)*num_children);
-    auto h_ptr = buffer.data();
+    // First calculate the size of memory needed to hold the
+    // child columns. This is done by calling extent()
+    // for each of the children.
+    size_type size_bytes = 0;
+    for( size_type idx=0; idx < num_children; ++idx )
+      size_bytes += extent(source.child(idx));
+    // A buffer of CPU memory is allocated to hold the column_device_view
+    // objects. Once filled, the CPU memory is then copied to device memory
+    // at the d_children member pointer.
+    std::vector<int8_t> h_buffer(size_bytes);
+    column_device_view* h_column = reinterpret_cast<column_device_view*>(h_buffer.data());
+    // Each column_device_view instance may have child objects that
+    // require setting some internal device pointers before being copied
+    // from CPU to device.
+    RMM_TRY(RMM_ALLOC(&p->d_children, size_bytes, stream));
```

Comment:
Use rmm::device_vector or rmm::device_buffer.

Reply:
The `p->d_children` here is a device memory pointer that must live as long as the `column_device_view` instance. It is freed as part of `destroy()` method that is registered with the returned `std::unique_ptr`.

Reply:
Got it.